src/dmd/root/array.d

```diff
@@ -803,10 +803,262 @@ if (is(ReturnType!(typeof((T t) => callable(t))) == int))
     assert(returnValue == 7);
 }
 
-private template ReturnType(T)
+/// Returns: A static array constructed from `array`.
+pragma(inline, true) T[n] staticArray(T, size_t n)(auto ref T[n] array)
+{
+    return array;
+}
+
+///
+pure nothrow @safe @nogc unittest
+{
+    enum a = [0, 1].staticArray;
+    static assert(is(typeof(a) == int[2]));
+    static assert(a == [0, 1]);
+}
+
+/// Returns: `true` if the two given ranges are equal
+bool equal(Range1, Range2)(Range1 range1, Range2 range2)
+{
+    template isArray (T)
+    {
+        static if (is(T U : U[]))
+            enum isArray = true;
+
+        else
+            enum isArray = false;
+    }
+
+    static if (isArray!Range1 && isArray!Range2 && is(typeof(range1 == range2)))
+        return range1 == range2;
+
+    else
+    {
+        static if (hasLength!Range1 && hasLength!Range2 && is(typeof(r1.length == r2.length)))
+        {
+            if (range1.length != range2.length)
+                return false;
+        }
+
+        else
+        {
+            for (; !range1.empty; range1.popFront(), range2.popFront())
+            {
+                if (range2.empty)
+                    return false;
+
+                if (range1.front != range2.front)
+                    return false;
+            }
```

Comment:
Needs to check `range2` for emptiness.
Test: `equal([1, 2].map(x => x * 2), [1, 2, 3].map(x => x * 2));`

Reply:
Done.